### PR TITLE
sql: Implement `ALTER SCHEMA ... RENAME ...`

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR implements the ability to rename a schema.

Not all of our `create_sql` statements are normalized with an id, e.g. `[u5 as materialize.public.foo]`, and neither are column names, e.g. `SELECT public.foo.a FROM foo`. Because of this, to rename a schema we need update all of the items in a schema, as well as everything that uses these items. This PR takes exactly that approach, the core logic uses a visitor to walk through a statement and renamed all references of the old schema.

This feature is behind a flag, `enable_rename_schema` largely because I still need to add audit log events. This diff was getting large enough as-is, so I decided to do this in a follow up.

### Motivation

Progress towards https://github.com/MaterializeInc/materialize/issues/3680

### Tips for reviewer

I broke this PR into several different commits which can be reviewed independently:

1. Updates to the SQL parser
2. Plumbing the new statement through planning
3. **Logic for renaming a schema**, this is the most important commit
4. Additions to the Catalog consistency checker
5. Tests
6. Adding the feature gate

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Adds the ability to do `ALTER SCHEMA ... RENAME TO ...`, behind a LaunchDarkly feature flag.

### CODE REVIEW

File: src/sql/src/rbac.rs [hunk 1]
```diff
@@ -909,6 +909,13 @@ fn generate_rbac_requirements(
             ownership: vec![ObjectId::Item(*id_a), ObjectId::Item(*id_b)],
             ..Default::default()
         },
+        Plan::AlterSchemaRename(plan::AlterSchemaRenamePlan {
+            cur_schema_spec,
+            new_schema_name: _,
+        }) => RbacRequirements {
+            ownership: vec![ObjectId::Schema(*cur_schema_spec)],
+            ..Default::default()
```
Comment (jkosh44) [inline src/sql/src/rbac.rs:917]: From the PostgreSQL docs (https://www.postgresql.org/docs/16/sql-alterschema.html)
> You must own the schema to use `ALTER SCHEMA`. To rename a schema you must also have the `CREATE` privilege for the database.

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -4941,12 +4942,61 @@ pub fn plan_alter_object_rename(
...
+    let Some((db_spec, schema_spec)) = resolve_schema(scx, name.clone(), if_exists)? else {
+        let object_type = ObjectType::Schema;
+        scx.catalog.add_notice(PlanNotice::ObjectDoesNotExist {
+            name: name.to_ast_string(),
+            object_type,
+        });
+        return Ok(Plan::AlterNoop(AlterNoopPlan { object_type }));
+    };
+
+    // TODO(parkmycar): Make this a structured error and re-evaluate if it needs to be an error.
```
Comment (jkosh44) [inline src/sql/src/plan/statement/ddl.rs:4971]: This is probably the right error to use: https://github.com/MaterializeInc/materialize/blob/100eacd09d512bf97fca847b90536ef4803785a5/src/adapter/src/catalog/error.rs#L45-L46

File: src/sql/src/plan/statement/ddl.rs [hunk 3]
```diff
@@ -4941,12 +4942,61 @@ pub fn plan_alter_object_rename(
...
+    };
+
+    // Make sure the name is unique.
+    let schemas = scx.catalog.get_database(db_id).schema_ids();
+    let new_schema_name = normalize::ident(to_schema_name);
+    if schemas.contains_key(&new_schema_name) {
+        return Err(PlanError::ObjectAlreadyExists {
+            name: new_schema_name,
+            object_type: ObjectType::Schema,
+        });
```
Comment (jkosh44) [inline src/sql/src/plan/statement/ddl.rs:4983]: Not a big deal for this PR, but it looks like this error has gotten a bit duplicated: https://github.com/MaterializeInc/materialize/blob/100eacd09d512bf97fca847b90536ef4803785a5/src/sql/src/catalog.rs#L1136-L1137

File: src/sql/src/plan/statement/ddl.rs [hunk 4]
```diff
@@ -4941,12 +4942,61 @@ pub fn plan_alter_object_rename(
...
+    }
+
+    // Prevent users from renaming system related schemas.
+    let schema = scx.catalog.get_schema(&db_spec, &schema_spec);
+    if schema.owner_id() == MZ_SYSTEM_ROLE_ID {
+        return Err(PlanError::NeverSupported {
+            feature: format!("renaming the {} schema", schema.name().schema),
+            documentation_link: None,
+            details: None,
+        });
```
Comment (jkosh44) [inline src/sql/src/plan/statement/ddl.rs:4993]: The `bail_never_supported!` macro might be handy here. It's subjective though, so up to you.

File: src/adapter/src/catalog/consistency.rs [hunk 5]
```diff
@@ -348,6 +358,143 @@ impl CatalogState {
             Err(dependency_inconsistencies)
         }
     }
+
```
Comment (jkosh44) [inline src/adapter/src/catalog/consistency.rs:361]: You could probably pull this whole commit out into a separate PR.

File: src/sql/src/rbac.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql/src/rbac.rs:917]: You're missing the `CREATE` privilege here.

**Review by jkosh44 — COMMENTED** [review]
I'm going to finish reviewing tomorrow, but here are some initial comments.

File: src/adapter/src/catalog.rs [hunk 6]
```diff
@@ -2849,6 +2850,114 @@ impl Catalog {
...
                     }
                 }
+                Op::RenameSchema {
+                    database_spec,
+                    schema_spec,
+                    new_name,
+                } => {
+                    let ResolvedDatabaseSpecifier::Id(database_id) = &database_spec else {
+                        let err = anyhow::anyhow!("Cannot rename schemas in the ambient database");
+                        return Err(AdapterError::Unstructured(err));
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:2860]: We should use the `ReadOnlySchema` error here. You can see an example in `Op::AlterSource`.

File: src/adapter/src/catalog.rs [hunk 7]
```diff
@@ -2849,6 +2850,114 @@ impl Catalog {
...
+                    };
+                    let SchemaSpecifier::Id(schema_id) = &schema_spec else {
+                        let err = anyhow::anyhow!("Cannot rename temporary schemas");
+                        return Err(AdapterError::Unstructured(err));
+                    };
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:2865]: Why is this an error? Either way I'd prefer to use a structured error here.

File: src/adapter/src/catalog.rs [hunk 8]
```diff
@@ -2849,6 +2850,114 @@ impl Catalog {
...
+
+                    let database = state.get_database(database_id);
+                    let schema = database.schemas_by_id.get(schema_id).ok_or_else(|| {
+                        let err = anyhow::anyhow!("unknown schema");
+                        AdapterError::Unstructured(err)
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:2870]: This seems like it's unreachable. If you were previously able to resolve the schema in the database then it must still exist. Still, I think we should use a structured error here. `mz_sql::catalog::CatalogError::UnknownSchema` seems appropriate.

File: src/adapter/src/catalog.rs [hunk 9]
```diff
@@ -2849,6 +2850,114 @@ impl Catalog {
...
+                        &database_spec,
+                        schema_id,
+                        -1,
+                    ));
+
+                    // Rename the schema itself.
+                    let database = state.get_database_mut(database_id);
+                    let schema = database.schemas_by_id.get_mut(schema_id).ok_or_else(|| {
+                        let err = anyhow::anyhow!("unknown schema");
+                        AdapterError::Unstructured(err)
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:2926]: Ditto

> This seems like it's unreachable. If you were previously able to resolve the schema in the database then it must still exist. Still, I think we should use a structured error here. mz_sql::catalog::CatalogError::UnknownSchema seems appropriate.

File: src/adapter/src/catalog/objects.rs [hunk 10]
```diff
@@ -860,6 +860,80 @@ impl CatalogItem {
...
+                i.create_sql = do_rewrite(i.create_sql);
+                Ok(CatalogItem::Connection(i))
+            }
+            CatalogItem::Type(i) => {
+                let mut i = i.clone();
+                i.create_sql = do_rewrite(i.create_sql);
+                Ok(CatalogItem::Type(i))
+            }
+            CatalogItem::Func(_) => {
+                unreachable!("{}s do not reference a schema", self.typ())
```
Comment (jkosh44) [inline src/adapter/src/catalog/objects.rs:932]: Why not `Ok(CatalogItem::Func(i.clone())),` like logs? Functions are still namespaced by a schema so it seems like in theory we could hit this branch if anyone tried to rename a schema that contained a function. In practice we never will because functions are only in system schemas and we don't (yet) allow creating functions in other schemas.

File: src/adapter/src/catalog/consistency.rs [hunk 11]
```diff
@@ -408,3 +555,42 @@ enum ObjectDependencyInconsistency {
...
     },
 }
+
+#[derive(Debug, Serialize, Clone, PartialEq)]
+enum ItemInconsistency {
+    /// The name in a `Database` `schemas_by_name` does not match the name on the `Schema` struct.
+    KeyedName {
+        db_schema_by_name: String,
+        struct_name: String,
+    },
```
Comment (jkosh44) [inline src/adapter/src/catalog/consistency.rs:565]: We should add similar checks for any of the following that don't already exist (Not necessarily in this PR):
- `database_by_name` and `database_by_id`
- `ambient_schemas_by_name` and `ambient_schemas_by_id`
- `clusters_by_name` and `clusters_by_id`
- `roles_by_name` and `roles_by_name`

File: src/adapter/src/catalog/consistency.rs [hunk 12]
```diff
@@ -408,3 +555,42 @@ enum ObjectDependencyInconsistency {
         object_b: GlobalId,
     },
 }
+
+#[derive(Debug, Serialize, Clone, PartialEq)]
+enum ItemInconsistency {
```
Comment (jkosh44) [inline src/adapter/src/catalog/consistency.rs:560]: Some of these inconsistencies seem specific to schemas or databases, not items.

File: src/sql/src/ast/transform.rs [hunk 13]
```diff
@@ -13,18 +13,121 @@ use std::collections::{BTreeMap, BTreeSet};
...
+    create_stmt: &mut Statement<Raw>,
+    cur_schema_name: &str,
+    new_schema_name: String,
+) {
+    let maybe_update_item_name = |item_name: &mut UnresolvedItemName| {
+        if let [.., schema, _item] = &mut item_name.0[..] {
+            if schema.as_str() == cur_schema_name {
+                *schema = Ident::new(new_schema_name.clone());
+            }
+        }
```
Comment (jkosh44) [inline src/sql/src/ast/transform.rs:39]: Should the else branch here be an error? When would a name ever not be fully qualified?

File: src/sql/src/ast/transform.rs [hunk 14]
```diff
@@ -13,18 +13,121 @@ use std::collections::{BTreeMap, BTreeSet};
...
+}
+
+impl<'a, 'ast> VisitMut<'ast, Raw> for CreateSqlRewriteSchema<'a> {
+    fn visit_expr_mut(&mut self, e: &'ast mut Expr<Raw>) {
+        match e {
+            Expr::Identifier(id) => {
+                // The last ID component is a column name that should not be
+                // considered in the rewrite.
+                let i = id.len() - 1;
+                self.maybe_rewrite_idents(&mut id[..i]);
```
Comment (jkosh44) [inline src/sql/src/ast/transform.rs:107]: It's been awhile since I've looked at this, how do we know it's always a column?

File: test/sqllogictest/rename.slt
Comment (jkosh44) [inline test/sqllogictest/rename.slt:1]: I may have missed this but a couple tests that we should add are:
- Making sure fully qualified column names in views are renamed.
- Create two schemas `"a"` and `"A"`, run some subset of the following and make sure the names are properly normalized w.r.t. capitalization.
    - `ALTER SCHEMA A RENAME ...`
    - `ALTER SCHEMA a RENAME ...`
    - `ALTER SCHEMA "A" RENAME ...`
    - `ALTER SCHEMA "a" RENAME ...`
- Create two schemas `a` and `b` and create some item with the name `a.b.c` (`b` is the schema name). Run `ALTER SCHEMA a RENAME ...` and make sure the item isn't accidentally renamed.

**Comment by jkosh44** [comment]
Let's make sure to get some QA eyes on this since they may be able to spot some interesting edge cases.

**Review by def- — CHANGES_REQUESTED** [review]
I'll add some tests for this, enable it by default in CI and check coverage in https://buildkite.com/materialize/coverage/builds/267
I'll push my changes directly to this branch later today.

Edit: Just saw Joe's comment:

> Let's make sure to get some QA eyes on this since they may be able to spot some interesting edge cases.

This was flagged as high-risk since it changed lots of code, so here I am!

File: src/adapter/src/catalog.rs [hunk 15]
```diff
@@ -2849,6 +2850,114 @@ impl Catalog {
...
+                        // Update our item.
+                        let mut new_entry = entry.clone();
+                        new_entry.item = entry
+                            .item
+                            .rename_schema_refs(&cur_name, new_name.clone())
+                            .expect("no failures");
+
+                        // Update the Stash and Builtin Tables.
+                        if !new_entry.item().is_temporary() {
+                            tx.update_item(*id, new_entry.clone().into())?;
```
Comment (def-) [inline src/adapter/src/catalog.rs:2893]: uncovered

File: src/adapter/src/catalog.rs (hunk 9 above)
Comment (def-) [reply src/adapter/src/catalog.rs:2926]: Uncovered

File: src/adapter/src/catalog/consistency.rs [hunk 16]
```diff
@@ -58,11 +65,14 @@ impl CatalogState {
         if let Err(dependencies) = self.check_object_dependencies() {
             inconsistencies.object_dependencies = dependencies;
         }
+        if let Err(items) = self.check_items() {
+            inconsistencies.items = items;
```
Comment (def-) [inline src/adapter/src/catalog/consistency.rs:69]: Uncovered

File: src/adapter/src/catalog/consistency.rs [hunk 17]
```diff
@@ -348,6 +358,143 @@ impl CatalogState {
...
+    /// * The item from the parsed `create_sql` must be fully qualified.
+    ///
+    fn check_items(&self) -> Result<(), Vec<ItemInconsistency>> {
+        let mut item_inconsistencies = vec![];
+
+        for (db_id, db) in &self.database_by_id {
+            for (schema_name, schema_id) in &db.schemas_by_name {
+                // Make sure the schema themselves are consistent.
+                let Some(schema) = db.schemas_by_id.get(schema_id) else {
+                    item_inconsistencies.push(ItemInconsistency::MissingSchema {
```
Comment (def-) [inline src/adapter/src/catalog/consistency.rs:380]: Uncovered

File: src/adapter/src/catalog/consistency.rs [hunk 18]
```diff
@@ -348,6 +358,143 @@ impl CatalogState {
...
+                        db_id: *db_id,
+                        schema_name: schema_name.clone(),
+                    });
+                    continue;
+                };
+                if schema_name != &schema.name.schema {
+                    item_inconsistencies.push(ItemInconsistency::KeyedName {
```
Comment (def-) [inline src/adapter/src/catalog/consistency.rs:387]: uncovered

File: src/adapter/src/catalog/consistency.rs [hunk 19]
```diff
@@ -348,6 +358,143 @@ impl CatalogState {
...
+                        db_schema_by_name: schema_name.clone(),
+                        struct_name: schema.name.schema.clone(),
+                    });
+                }
+
+                // Make sure the items in the schema are consistent.
+                for (_, item_id) in &schema.items {
+                    let Some(entry) = self.entry_by_id.get(item_id) else {
+                        item_inconsistencies.push(ItemInconsistency::NonExistentItem {
```
Comment (def-) [inline src/adapter/src/catalog/consistency.rs:396]: uncovered

File: src/adapter/src/catalog/consistency.rs [hunk 20]
```diff
@@ -348,6 +358,143 @@ impl CatalogState {
...
+                        });
+                        continue;
+                    };
+                    let statement = match mz_sql::parse::parse(entry.create_sql()) {
+                        Ok(mut statements) if statements.len() == 1 => {
+                            let statement = statements.pop().expect("checked length");
+                            statement.ast
+                        }
+                        Ok(_) => {
+                            item_inconsistencies.push(ItemInconsistency::MultiCreateStatement {
```
Comment (def-) [inline src/adapter/src/catalog/consistency.rs:409]: uncovered

File: src/adapter/src/catalog/consistency.rs [hunk 21]
```diff
@@ -348,6 +358,143 @@ impl CatalogState {
...
+                                create_sql: entry.create_sql().to_string(),
+                            });
+                            continue;
+                        }
+                        Err(_) => {
+                            item_inconsistencies.push(ItemInconsistency::StatementParseFailure {
```
Comment (def-) [inline src/adapter/src/catalog/consistency.rs:415]: uncovered

File: src/sql/src/plan/statement/ddl.rs [hunk 22]
```diff
@@ -4941,12 +4942,63 @@ pub fn plan_alter_object_rename(
...
+pub fn plan_alter_schema_rename(
+    scx: &mut StatementContext,
+    name: UnresolvedSchemaName,
+    to_schema_name: Ident,
+    if_exists: bool,
+) -> Result<Plan, PlanError> {
+    scx.require_feature_flag(&vars::ENABLE_RENAME_SCHEMA)?;
+
+    let Some((db_spec, schema_spec)) = resolve_schema(scx, name.clone(), if_exists)? else {
+        let object_type = ObjectType::Schema;
```
Comment (def-) [inline src/sql/src/plan/statement/ddl.rs:4963]: uncovered

File: src/sql/src/ast/transform.rs [hunk 23]
```diff
@@ -13,18 +13,121 @@ use std::collections::{BTreeMap, BTreeSet};
...
+                self.maybe_rewrite_idents(id);
+            }
+            _ => visit_mut::visit_expr_mut(self, e),
+        }
+    }
+    fn visit_unresolved_item_name_mut(
+        &mut self,
+        unresolved_item_name: &'ast mut UnresolvedItemName,
+    ) {
+        self.maybe_rewrite_idents(&mut unresolved_item_name.0);
```
Comment (def-) [inline src/sql/src/ast/transform.rs:119]: uncovered

**Review by def- — COMMENTED** [review]
Somehow lots of interesting code is still untested: https://buildkite.com/materialize/coverage/builds/267#_ I marked it inline. Most should be coverable with explicit slt tests I guess?

**Comment by def-** [comment]
I pushed two kinds of tests for this (platform-check for upgrades/restarts & parallel-workload), had to rebase because of conflicts with main.

File: src/sql/src/rbac.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/sql/src/rbac.rs]: Updated! I didn't know what permission to use for schemas in the ambient database, so I kind of guessed.

File: src/sql/src/plan/statement/ddl.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs]: Refactored this away

File: src/sql/src/plan/statement/ddl.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs]: Ahhh I missed this error originally, I reverted the change to `ObjectAlreadyExists` and re-used the one you mentioned

File: src/sql/src/plan/statement/ddl.rs (hunk 4 above)
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs]: Good idea, updated!

File: src/adapter/src/catalog/consistency.rs (hunk 5 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: I pulled this our into https://github.com/MaterializeInc/materialize/pull/22411

File: src/adapter/src/catalog.rs (hunk 6 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog.rs]: Refactored this away

File: src/adapter/src/catalog.rs (hunk 7 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog.rs]: Good thought, I refactored this away

File: src/adapter/src/catalog.rs (hunk 8 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog.rs]: Refactored it away

File: src/adapter/src/catalog.rs (hunk 9 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog.rs]: Refactored it away

File: src/adapter/src/catalog/objects.rs (hunk 10 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/objects.rs]: Good idea! I updated this to return an `Ok(CatalogItem::Func(i.clone()))`

File: src/adapter/src/catalog/consistency.rs (hunk 11 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: FWIW these mostly are already covered [here](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/consistency.rs#L85-L124)

File: src/adapter/src/catalog/consistency.rs (hunk 12 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: Yeah... the naming wasn't the best. This was pulled out of this PR though

File: src/sql/src/ast/transform.rs (hunk 13 above)
Comment (ParkMyCar) [reply src/sql/src/ast/transform.rs:39]: Tbh I'm not sure. Currently since `UnresolvedItemName` is just a collection of idents, so there is no type-level guarantee this will be fully qualified. There is an issue (https://github.com/MaterializeInc/materialize/issues/5591) to express how well qualified an ident is.

If something is not fully qualified then it's probably okay? No need to rename anything then?

File: src/sql/src/ast/transform.rs (hunk 14 above)
Comment (ParkMyCar) [reply src/sql/src/ast/transform.rs:107]: It's just by convention, there is an issue (https://github.com/MaterializeInc/materialize/issues/5591) to better improve this for some type level guarantees.

File: test/sqllogictest/rename.slt
Comment (ParkMyCar) [reply test/sqllogictest/rename.slt:1]: Updated to cover these cases!

File: src/adapter/src/catalog/consistency.rs (hunk 16 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: Part of the consistency checker, pulled out of this PR

File: src/adapter/src/catalog/consistency.rs (hunk 18 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: Part of the consistency checker, pulled out of this PR

File: src/adapter/src/catalog/consistency.rs (hunk 19 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: Part of the consistency checker, pulled out of this PR

File: src/adapter/src/catalog/consistency.rs (hunk 20 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: Part of the consistency checker, pulled out of this PR

File: src/adapter/src/catalog/consistency.rs (hunk 21 above)
Comment (ParkMyCar) [reply src/adapter/src/catalog/consistency.rs]: Part of the consistency checker, pulled out of this PR

File: src/sql/src/plan/statement/ddl.rs (hunk 22 above)
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs:4963]: Should be covered now 🤞

File: src/sql/src/ast/transform.rs (hunk 23 above)
Comment (ParkMyCar) [reply src/sql/src/ast/transform.rs:119]: This is surprising! Not sure why this is uncovered

**Comment by ParkMyCar** [comment]
> I pushed two kinds of tests for this (platform-check for upgrades/restarts & parallel-workload), had to rebase because of conflicts with main.

Thanks for adding these tests @def-! I patched your commit on top of my latest push

File: src/sql/src/rbac.rs [hunk 24]
```diff
@@ -909,6 +909,27 @@ fn generate_rbac_requirements(
...
+            new_schema_name: _,
+        }) => {
+            let privileges = match cur_schema_spec.0 {
+                ResolvedDatabaseSpecifier::Id(db_id) => vec![(
+                    SystemObjectId::Object(ObjectId::Database(db_id)),
+                    AclMode::CREATE,
+                    role_id,
+                )],
+                ResolvedDatabaseSpecifier::Ambient => {
+                    vec![(SystemObjectId::System, AclMode::CREATE, role_id)]
```
Comment (jkosh44) [inline src/sql/src/rbac.rs:923]: This isn't quite correct. This is reachable for temporary schemas right? If so, we should not require any additional privileges here.

File: src/sql/src/ast/transform.rs (hunk 13 above)
Comment (jkosh44) [reply src/sql/src/ast/transform.rs:39]: > If something is not fully qualified then it's probably okay? No need to rename anything then?

Yeah that's a good point. I don't think we have to worry about it in this PR.

File: src/sql/src/plan/statement/ddl.rs [hunk 25]
```diff
@@ -4941,12 +4942,54 @@ pub fn plan_alter_object_rename(
...
+        return Err(PlanError::Catalog(CatalogError::SchemaAlreadyExists(
+            to_schema_name.clone().into_string(),
+        )));
+    }
+
+    // Prevent users from renaming system related schemas.
+    let schema = scx.catalog.get_schema(&db_spec, &schema_spec);
+    if schema.owner_id() == MZ_SYSTEM_ROLE_ID {
+        bail_never_supported!(format!("renaming the {} schema", schema.name().schema))
+    }
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs:4878]: I used `schema.owner_id()` because the "public" schema is a User schema, and I didn't want to allow folks to rename that, what do you think?

File: src/sql/src/rbac.rs (hunk 24 above)
Comment (ParkMyCar) [reply src/sql/src/rbac.rs]: Sweet, updated!

File: src/sql/src/plan/statement/ddl.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs:4878]: The `public` schema in general is not owned by `mz_system`

```
materialize=> CREATE DATABASE db;
CREATE DATABASE
materialize=> SELECT * FROM mz_schemas WHERE name = 'public';
 id |  oid  | database_id |  name  | owner_id |            privileges             
----+-------+-------------+--------+----------+-----------------------------------
 u6 | 20603 | u2          | public | u1       | {s2=U/u1,u1=UC/u1,=U/u1}
 u3 | 20006 | u1          | public | s1       | {s1=UC/s1,s2=U/s1,u1=UC/s1,=U/s1}
(2 rows)

```
(Note the owner id of `u1`).

It's just that the `public` schema in the `materialize` database is owned by `mz_system` because there's no one else to own it. In general we treat `public` schemas the same as every other schema, it's just automatically created with some sane defaults. So I think we should allow users to rename the `public` schema.

[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql/src/plan/statement/ddl.rs]: Makes sense! Updated to use `schema.id().is_system()`

**Comment by ParkMyCar** [comment]
@def- anything else you'd like to see from a QA point of view?

**Review by def- — APPROVED** [review]
Thanks!
